File: ios/WidgetExtensionTemplate/WidgetContentEngine.swift
import Foundation

struct HourlyWidgetContent {
    let hadithSource: String
    let hadithText: String
}

enum WidgetContentEngine {
    static func pickHourlyContent(date: Date = Date()) -> HourlyWidgetContent {
        let hadiths = loadArray(named: "widget_hadiths")

        guard !hadiths.isEmpty else {
            return HourlyWidgetContent(
                hadithSource: "Hadis",
                hadithText: "Saatlik içerik yüklenemedi."
            )
        }

        let seed = hourlySeed(for: date)
        let hadithIndex = Int(floor(seededRandom(seed * 7 + 13) * Double(hadiths.count)))
        let hadith = hadiths[hadithIndex]

        return HourlyWidgetContent(
            hadithSource: hadith["s"] as? String ?? "Hadis",
            hadithText: hadith["t"] as? String ?? ""
        )
    }

    private static func loadArray(named: String) -> [[String: Any]] {
        guard let url = Bundle.main.url(forResource: named, withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let raw = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }
        return raw
    }

    private static func hourlySeed(for date: Date) -> Int {
        let c = Calendar.current.dateComponents([.year, .month, .day, .hour], from: date)
        let year = c.year ?? 0
        let month = c.month ?? 0
        let day = c.day ?? 0
        let hour = c.hour ?? 0
        return year * 1_000_000 + month * 10_000 + day * 100 + hour
    }

    private static func seededRandom(_ seed: Int) -> Double {
        let x = sin(Double(seed)) * 10000
        return x - floor(x)
    }
}

File: android/app/src/main/java/com/hikmetehli/app/WidgetContentEngine.java
package com.hikmetehli.app;

import android.content.Context;
import android.content.res.AssetManager;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.Calendar;

public final class WidgetContentEngine {
    private static JSONArray hadithCache;

    private WidgetContentEngine() {}

    public static HourlyContent getHourlyContent(Context context) {
        ensureLoaded(context);
        int seed = getHourlySeed();

        if (hadithCache == null || hadithCache.length() == 0) {
            return HourlyContent.fallback();
        }

        int hadithIndex = (int) Math.floor(seededRandom(seed * 7 + 13) * hadithCache.length());

        try {
            JSONObject hadith = hadithCache.getJSONObject(hadithIndex);
            return new HourlyContent(
                hadith.optString("t", "Hadis yüklenemedi."),
                hadith.optString("s", ""));
        } catch (Exception ignored) {
            return HourlyContent.fallback();
        }
    }

    private static void ensureLoaded(Context context) {
        if (hadithCache != null) {
            return;
        }
        hadithCache = loadJsonArray(context.getAssets(), "widget_hadiths.json");
    }

    private static JSONArray loadJsonArray(AssetManager assets, String assetName) {
        try {
            InputStream stream = assets.open(assetName);
            BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8));
            StringBuilder builder = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                builder.append(line);
            }
            reader.close();
            return new JSONArray(builder.toString());
        } catch (Exception ignored) {
            return null;
        }
    }

    private static int getHourlySeed() {
        Calendar now = Calendar.getInstance();
        return now.get(Calendar.YEAR) * 1000000
            + (now.get(Calendar.MONTH) + 1) * 10000
            + now.get(Calendar.DAY_OF_MONTH) * 100
            + now.get(Calendar.HOUR_OF_DAY);
    }

    private static double seededRandom(int seed) {
        double x = Math.sin(seed) * 10000;
        return x - Math.floor(x);
    }

    public static class HourlyContent {
        public final String hadithText;
        public final String hadithSource;

        public HourlyContent(
            String hadithText,
            String hadithSource
        ) {
            this.hadithText = hadithText;
            this.hadithSource = hadithSource;
        }

        static HourlyContent fallback() {
            return new HourlyContent(
                "Saatlik hadis içeriği yüklenemedi.",
                ""
            );
        }
    }
}
